src/L2/UpgradeableRegistrarController.sol

```diff
@@ -0,0 +1,676 @@
+            return $.launchTime;
+        }
+        return expires + GRACE_PERIOD;
+    }
+
+    /// @notice Shared registartion logic for both `register()` and `discountedRegister()`.
+    ///
+    /// @dev Will set records in the specified resolver if the resolver address is non zero and there is `data` in the `request`.
+    ///     Will set the reverse record's owner as msg.sender if `reverseRecord` is `true`.
+    ///     Emits `NameRegistered` upon successful registration.
+    ///
+    /// @param request The `RegisterRequest` struct containing the details for the registration.
+    function _register(RegisterRequest calldata request) internal {
+        uint256 expires = _getURCStorage().base.registerWithRecord(
+            uint256(keccak256(bytes(request.name))), request.owner, request.duration, request.resolver, 0
+        );
+
+        if (request.data.length > 0) {
+            _setRecords(request.resolver, keccak256(bytes(request.name)), request.data);
+        }
+
+        if (request.reverseRecord) {
+            _setReverseRecord(request.name, request.resolver, msg.sender, request.signatureExpiry, request.signature);
+        }
+
+        emit NameRegistered(request.name, keccak256(bytes(request.name)), request.owner, expires);
+    }
+
+    /// @notice Refunds any remaining `msg.value` after processing a registration or renewal given`price`.
+    ///
+    /// @dev It is necessary to allow "overpayment" because of premium price decay.  We don't want transactions to fail
+    ///     unnecessarily if the premium decreases between tx submission and inclusion.
+    ///
+    /// @param price The total value to be retained, denominated in wei.
+    function _refundExcessEth(uint256 price) internal {
+        if (msg.value > price) {
+            (bool sent,) = payable(msg.sender).call{value: (msg.value - price)}("");
+            if (!sent) revert TransferFailed();
+        }
+    }
+
+    /// @notice Uses Multicallable to iteratively set records on a specified resolver.
+    ///
+    /// @dev `multicallWithNodeCheck` ensures that each record being set is for the specified `label`.
+    ///
+    /// @param resolverAddress The address of the resolver to set records on.
+    /// @param label The keccak256 namehash for the specified name.
+    /// @param data  The abi encoded calldata records that will be used in the multicallable resolver.
+    function _setRecords(address resolverAddress, bytes32 label, bytes[] calldata data) internal {
+        bytes32 nodehash = keccak256(abi.encodePacked(_getURCStorage().rootNode, label));
+        L2Resolver resolver = L2Resolver(resolverAddress);
+        resolver.multicallWithNodeCheck(nodehash, data);
+    }
+
+    /// @notice Sets the reverse record to `owner` for a specified `name` on the specified `resolver.
+    ///
+    /// @param name The specified name.
+    /// @param resolver The resolver to set the reverse record on.
+    /// @param owner  The owner of the reverse record.
```

Comment:
Missing natspec for params:
        uint256 expiry,
        bytes memory signature